src/algebra/quotient.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2021 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+
+import data.set_like.basic
+
+/-!
+# Algebraic quotients
+
+This file defines notation for algebraic quotients, e.g. quotient groups `G /// H`,
+quotient modules `M /// N` and ideal quotients `R /// I`.
+
+The actual quotient structures are defined in the following files:
+ * quotient group: `src/group_theory/quotient_group.lean`
+ * quotient module: `src/linear_algebra/quotient.lean`
+ * quotient ring: `src/ring_theory/ideal/quotient.lean`
+
+## Notations
+
+The following notation is introduced:
+
+* `G /// H` stands for the quotient of the type `G` by some term `H` 
+  (for example, `H` can be a normal subgroup of `G`).
+  To implement this notation for other quotients, you should provide a `has_quotient` instance.
+  Note that since `G` can usually be inferred from `H`, `_ /// H` can also be used,
+  but this is less readable.
+
+## Tags
+
+quotient, group quotient, quotient group, module quotient, quotient module, ring quotient,
+ideal quotient, quotient ring
+-/
+
+universes u v
+
+/-- `has_quotient A B` is a notation typeclass that allows us to write `A /// b` for `b : B`.
+This allows the usual notation for quotients of algebraic structures,
+such as groups, modules and rings.
+
+`A` is a parameter, despite being unused in the definition below, so it appears in the notation.
+-/
+class has_quotient (A : out_param $ Type u) (B : Type v) :=
+(quotient' : B → Type (max u v))
+
+/-- `has_quotient.quotient A b` (with notation `A /// b`) is the quotient of the type `A` by `b`.
+
+This differs from `has_quotient.quotient'` in that the `A` argument is explicit, which is necessary
```

Comment:
Can't you get the explicitness by adding some `[]` or `{}` after `quotient'` in the def'n above?
cc @gebner

Reply:
I tried that and got nowhere

Reply:
Indeed, I tried all the brackets I could think of before just giving up and making a new definition.

Reply:
The square `[]` should've worked.  And it does if you removed the `out_param` :shrug:.

Reply:
The current approach, with `out_param`, looks the best notation-wise to me. So I'd like to keep that.